lib/ASTGen/Sources/MacroEvaluation/Macros.swift

```diff
@@ -583,15 +583,22 @@ func expandAttachedMacro(
     return 1
   }
 
-  // Dig out the node for the declaration to which the custom attribute is
-  // attached.
-  guard
-    let declarationNode = findSyntaxNodeInSourceFile(
+  func findNode<T: SyntaxProtocol>(type: T.Type) -> T? {
+    findSyntaxNodeInSourceFile(
       sourceFilePtr: declarationSourceFilePtr,
       sourceLocationPtr: declarationSourceLocPointer,
-      type: DeclSyntax.self
+      type: T.self
     )
-  else {
+  }
+
+  // Dig out the node for the closure or declaration to which the custom
+  // attribute is attached.
+  let node: Syntax
+  if let closureNode = findNode(type: ClosureExprSyntax.self) {
+    node = Syntax(closureNode)
```

Comment:
This may find outer unrelated `ClosureSyntax`. E.g.
```swift
_ = { (arg: Int) in
  @myBodyMacro func foo() -> Void
}
```
In this case the source location would be the `func` which is inside the closure. We need additional check, or maybe `findSyntaxNodeInSourceFile` should accept a check function like:

```swift
findSyntaxNodeInSourceFile(
  sourceFilePtr: declarationSourceFilePtr,
  sourceLocationPtr: declarationSourceLocPointer,
  where: { $0.is(DeclSyntax.self) || $0.is(ClosureExprSyntax.self) }
)

Reply:
Good catch, thank you! I like your idea to add an overload of `findSyntaxNodeInSourceFile` that accepts a predicate, I'll make that change.